engine/table/src/main/java/io/deephaven/engine/context/QueryCompilerImpl.java

```diff
@@ -0,0 +1,1037 @@
+            this.description = description;
+            this.code = code;
+            this.resolver = resolver;
+        }
+
+        public CharSequence getCharContent(boolean ignoreEncodingErrors) {
+            return code;
+        }
+    }
+
+    private static class CompilationRequestAttempt {
+        final String description;
+        final String fqClassName;
+        final String finalCode;
+        final String packageName;
+        final String[] splitPackageName;
+        final QueryCompilerRequest request;
+        final CompletionStageFuture.Resolver<Class<?>> resolver;
+
+        private CompilationRequestAttempt(
+                @NotNull final QueryCompilerRequest request,
+                @NotNull final String packageName,
+                @NotNull final String fqClassName,
+                @NotNull final CompletionStageFuture.Resolver<Class<?>> resolver) {
+            this.description = request.description();
+            this.fqClassName = fqClassName;
+            this.resolver = resolver;
+            this.packageName = packageName;
+            this.request = request;
+
+            finalCode = makeFinalCode(request.className(), request.classBody(), packageName);
+
+            if (logEnabled) {
+                log.info().append("Generating code ").append(finalCode).endl();
+            }
+
+            splitPackageName = packageName.split("\\.");
+            if (splitPackageName.length == 0) {
+                final Exception err = new UncheckedDeephavenException(String.format(
+                        "packageName %s expected to have at least one .", packageName));
+                resolver.completeExceptionally(err);
+            }
+        }
+
+        public void ensureDirectories(@NotNull final String rootPath) {
+            if (splitPackageName.length == 0) {
+                // we've already failed
+                return;
+            }
+
+            final String[] truncatedSplitPackageName = Arrays.copyOf(splitPackageName, splitPackageName.length - 1);
+            final Path rootPathWithPackage = Paths.get(rootPath, truncatedSplitPackageName);
+            final File rpf = rootPathWithPackage.toFile();
+            QueryCompilerImpl.ensureDirectories(rpf,
+                    () -> "Couldn't create package directories: " + rootPathWithPackage);
+        }
+
+        public JavaSourceFromString makeSource() {
+            return new JavaSourceFromString(description, fqClassName, finalCode, resolver);
+        }
+    }
+
+    private static final JavaCompiler compiler;
+    private static final JavaFileManager fileManager;
+    static {
+        compiler = ToolProvider.getSystemJavaCompiler();
+        if (compiler == null) {
+            throw new UncheckedDeephavenException("No Java compiler provided - are you using a JRE instead of a JDK?");
+        }
+        fileManager = new SynchronizedJavaFileManager(compiler.getStandardFileManager(null, null, null));
+    }
```

Comment:
After initializing in the constructor the thrown error looks good:
```
Initiating shutdown due to: Uncaught exception in thread main
io.deephaven.UncheckedDeephavenException: No Java compiler provided - are you using a JRE instead of a JDK?
	at io.deephaven.engine.context.QueryCompilerImpl.ensureJavaCompiler(QueryCompilerImpl.java:83)
	at io.deephaven.engine.context.QueryCompilerImpl.<init>(QueryCompilerImpl.java:131)
	at io.deephaven.engine.context.QueryCompilerImpl.create(QueryCompilerImpl.java:110)
	at io.deephaven.engine.util.AbstractScriptSession.<init>(AbstractScriptSession.java:98)
	at io.deephaven.engine.util.AbstractScriptSession.<init>(AbstractScriptSession.java:81)
	at io.deephaven.integrations.python.PythonDeephavenSession.<init>(PythonDeephavenSession.java:86)
	at io.deephaven.server.console.python.PythonConsoleSessionModule.bindPythonSession(PythonConsoleSessionModule.java:41)
	at io.deephaven.server.console.python.PythonConsoleSessionModule_BindPythonSessionFactory.bindPythonSession(PythonConsoleSessionModule_BindPythonSessionFactory.java:77)
	at io.deephaven.server.console.python.PythonConsoleSessionModule_BindPythonSessionFactory.get(PythonConsoleSessionModule_BindPythonSessionFactory.java:61)
	at io.deephaven.server.console.python.PythonConsoleSessionModule_BindPythonSessionFactory.get(PythonConsoleSessionModule_BindPythonSessionFactory.java:18)
	at io.deephaven.server.console.python.PythonConsoleSessionModule_BindScriptSessionFactory.get(PythonConsoleSessionModule_BindScriptSessionFactory.java:37)
	at io.deephaven.server.console.python.PythonConsoleSessionModule_BindScriptSessionFactory.get(PythonConsoleSessionModule_BindScriptSessionFactory.java:13)
	at io.deephaven.server.session.ScriptSessionModule.provideScriptSession(ScriptSessionModule.java:31)
	at io.deephaven.server.session.ScriptSessionModule_ProvideScriptSessionFactory.provideScriptSession(ScriptSessionModule_ProvideScriptSessionFactory.java:44)
	at io.deephaven.server.session.ScriptSessionModule_ProvideScriptSessionFactory.get(ScriptSessionModule_ProvideScriptSessionFactory.java:34)
	at io.deephaven.server.session.ScriptSessionModule_ProvideScriptSessionFactory.get(ScriptSessionModule_ProvideScriptSessionFactory.java:13)
	at dagger.internal.DoubleCheck.get(DoubleCheck.java:47)
	at io.deephaven.server.runner.DeephavenApiServer.run(DeephavenApiServer.java:151)
	at io.deephaven.server.jetty.JettyMain.main(JettyMain.java:24)
```